### PR TITLE
Create a worker for event creation

### PR DESCRIPTION
**Pull request by erikjohnston**
Includes ~~https://github.com/matrix-org/synapse/pull/2847~~ https://github.com/matrix-org/synapse/pull/2856

### CODE REVIEW

**Comment by turt2live** [comment]
Please add documentation for this worker to https://github.com/matrix-org/synapse/blob/develop/docs/workers.rst (including the routes that need to be redirected to it).

File: synapse/app/homeserver.py [hunk 1]
```diff
@@ -219,6 +220,9 @@ def _configure_named_resource(self, name, compress=False):
         if name == "metrics" and self.get_config().enable_metrics:
             resources[METRICS_PREFIX] = MetricsResource(self)
 
+        if name == "replication":
```
Comment (richvdh) [inline synapse/app/homeserver.py]: needs documenting?

File: docs/workers.rst [hunk 2]
```diff
@@ -207,3 +207,14 @@ the ``worker_main_http_uri`` setting in the frontend_proxy worker configuration
...
 
     worker_main_http_uri: http://127.0.0.1:8008
+
+
+``synapse.app.event_creator``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Handles non-state event creation. It can handle REST endpoints matching:
+
+    ^/_matrix/client/(api/v1|r0|unstable)/rooms/.*/send
```
Comment (ara4n) [inline docs/workers.rst]: Shouldn't this also be handling the macro endpoints for creating events (https://matrix.org/docs/spec/client_server/r0.3.0.html#post-matrix-client-r0-rooms-roomid-ban etc)? Or are they deliberately being left on the master as they're not that common?

Comment (erikjohnston) [reply docs/workers.rst]: We're currently only handling non-state events on this worker for now, as sometimes the membership state changes end up doing complicated things that probably need to be done on the master.

**Comment by erikjohnston** [comment]
> this generally looks sane, but it feels to me like the "http replication" we're introducing here is quite orthogonal to the existing TCP replication (and the old HTTP replication we used to have), and isn't really replication at all.

I've taken replication to mean all internal traffic between two nodes, both downstream and upstream communications. I can changes the name I guess, but it means internal communication is now under two names rather than one.

File: docs/workers.rst [hunk 3]
```diff
@@ -30,17 +30,29 @@ requests made to the federation port. The caveats regarding running a
...
+         - names: [replication]
 
-Under **no circumstances** should this replication API listener be exposed to the
-public internet; it currently implements no authentication whatsoever and is
+Under **no circumstances** should these replication API listeners be exposed to
+the public internet; it currently implements no authentication whatsoever and is
 unencrypted.
 
+(Roughly, the TCP port is used for streaming data from the master to the
+workers, and the HTTP port for the workers to communicate with the main
```
Comment (richvdh) [inline docs/workers.rst]: s/communicate with/send data to/

File: docs/workers.rst [hunk 4]
```diff
@@ -52,8 +64,10 @@ You should minimise the number of overrides though to maintain a usable config.
 
 You must specify the type of worker application (``worker_app``). The currently
 available worker applications are listed below. You must also specify the
-replication endpoint that it's talking to on the main synapse process
-(``worker_replication_host`` and ``worker_replication_port``).
+replication endpoints that it's talking to on the main synapse process.
+``worker_replication_host`` should specify the host of the main synapse,
+``worker_replication_port`` should point to the TCP replication listener port and
+``worker_replication_http_port`` should point to the HTTP replication port.
```
Comment (richvdh) [inline docs/workers.rst]: might be worth mentioning here that only the event_creator worker needs this for now, to help people upgrading.

**Pull request merged**